File: examples/Algorithm/c/04_hash_table.c
/*
 * Hash Table
 * Hash Functions, Chaining, Open Addressing
 *
 * A data structure for fast search, insertion, and deletion.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define TABLE_SIZE 101
#define DELETED ((void*)-1)

/* =============================================================================
 * 1. Hash Functions
 * ============================================================================= */

/* Modular hash */
unsigned int hash_mod(int key, int size) {
    return ((key % size) + size) % size;
}

/* String hash (djb2) */
unsigned int hash_string(const char* str, int size) {
    unsigned long hash = 5381;
    int c;
    while ((c = *str++)) {
        hash = ((hash << 5) + hash) + c;
    }
    return hash % size;
}

/* Polynomial hash */
unsigned int hash_polynomial(const char* str, int size) {
    unsigned long hash = 0;
    unsigned long p_pow = 1;
    const unsigned long p = 31;
    while (*str) {
        hash = (hash + (*str - 'a' + 1) * p_pow) % size;
        p_pow = (p_pow * p) % size;
        str++;
    }
    return hash;
}

/* =============================================================================
 * 2. Chaining Hash Table
 * ============================================================================= */

typedef struct ChainNode {
    char* key;
    int value;
    struct ChainNode* next;
} ChainNode;

typedef struct {
    ChainNode** buckets;
    int size;
    int count;
} ChainHashTable;

ChainHashTable* chain_create(int size) {
    ChainHashTable* ht = malloc(sizeof(ChainHashTable));
    ht->buckets = calloc(size, sizeof(ChainNode*));
    ht->size = size;
    ht->count = 0;
    return ht;
}

void chain_insert(ChainHashTable* ht, const char* key, int value) {
    unsigned int idx = hash_string(key, ht->size);

    /* Search for existing key */
    ChainNode* node = ht->buckets[idx];
    while (node) {
        if (strcmp(node->key, key) == 0) {
            node->value = value;
            return;
        }
        node = node->next;
    }

    /* Insert new node */
    ChainNode* new_node = malloc(sizeof(ChainNode));
    new_node->key = strdup(key);
    new_node->value = value;
    new_node->next = ht->buckets[idx];
    ht->buckets[idx] = new_node;
    ht->count++;
}

int chain_get(ChainHashTable* ht, const char* key, int* found) {
    unsigned int idx = hash_string(key, ht->size);
    ChainNode* node = ht->buckets[idx];

    while (node) {
        if (strcmp(node->key, key) == 0) {
            *found = 1;
            return node->value;
        }
        node = node->next;
    }

    *found = 0;
    return 0;
}

void chain_delete(ChainHashTable* ht, const char* key) {
    unsigned int idx = hash_string(key, ht->size);
    ChainNode* node = ht->buckets[idx];
    ChainNode* prev = NULL;

    while (node) {
        if (strcmp(node->key, key) == 0) {
            if (prev) {
                prev->next = node->next;
            } else {
                ht->buckets[idx] = node->next;
            }
            free(node->key);
            free(node);
            ht->count--;
            return;
        }
        prev = node;
        node = node->next;
    }
}

void chain_free(ChainHashTable* ht) {
    for (int i = 0; i < ht->size; i++) {
        ChainNode* node = ht->buckets[i];
        while (node) {
            ChainNode* next = node->next;
            free(node->key);
            free(node);
            node = next;
        }
    }
    free(ht->buckets);
    free(ht);
}

/* =============================================================================
 * 3. Open Addressing (Linear Probing)
 * ============================================================================= */

typedef struct {
    int* keys;
    int* values;
    bool* occupied;
    bool* deleted;
    int size;
    int count;
} LinearHashTable;

LinearHashTable* linear_create(int size) {
    LinearHashTable* ht = malloc(sizeof(LinearHashTable));
    ht->keys = malloc(size * sizeof(int));
    ht->values = malloc(size * sizeof(int));
    ht->occupied = calloc(size, sizeof(bool));
    ht->deleted = calloc(size, sizeof(bool));
    ht->size = size;
    ht->count = 0;
    return ht;
}

void linear_insert(LinearHashTable* ht, int key, int value) {
    if (ht->count >= ht->size * 0.7) {
        printf("    Warning: Load factor too high!\n");
        return;
    }

    unsigned int idx = hash_mod(key, ht->size);

    while (ht->occupied[idx] && !ht->deleted[idx] && ht->keys[idx] != key) {
        idx = (idx + 1) % ht->size;
    }

    if (!ht->occupied[idx] || ht->deleted[idx]) {
        ht->count++;
    }

    ht->keys[idx] = key;
    ht->values[idx] = value;
    ht->occupied[idx] = true;
    ht->deleted[idx] = false;
}

int linear_get(LinearHashTable* ht, int key, int* found) {
    unsigned int idx = hash_mod(key, ht->size);
    int start = idx;

    while (ht->occupied[idx]) {
        if (!ht->deleted[idx] && ht->keys[idx] == key) {
            *found = 1;
            return ht->values[idx];
        }
        idx = (idx + 1) % ht->size;
        if (idx == start) break;
    }

    *found = 0;
    return 0;
}

void linear_delete(LinearHashTable* ht, int key) {
    unsigned int idx = hash_mod(key, ht->size);
    int start = idx;

    while (ht->occupied[idx]) {
        if (!ht->deleted[idx] && ht->keys[idx] == key) {
            ht->deleted[idx] = true;
            ht->count--;
            return;
        }
        idx = (idx + 1) % ht->size;
        if (idx == start) break;
    }
}

void linear_free(LinearHashTable* ht) {
    free(ht->keys);
    free(ht->values);
    free(ht->occupied);
    free(ht->deleted);
    free(ht);
}

/* =============================================================================
 * 4. Double Hashing
 * ============================================================================= */

typedef struct {
    int* keys;
    int* values;
    bool* occupied;
    bool* deleted;
    int size;
    int count;
} DoubleHashTable;

unsigned int hash2(int key, int size) {
    return 7 - (key % 7);  /* Ensures non-zero value */
}

DoubleHashTable* double_create(int size) {
    DoubleHashTable* ht = malloc(sizeof(DoubleHashTable));
    ht->keys = malloc(size * sizeof(int));
    ht->values = malloc(size * sizeof(int));
    ht->occupied = calloc(size, sizeof(bool));
    ht->deleted = calloc(size, sizeof(bool));
    ht->size = size;
    ht->count = 0;
    return ht;
}

void double_insert(DoubleHashTable* ht, int key, int value) {
    unsigned int idx = hash_mod(key, ht->size);
    unsigned int step = hash2(key, ht->size);

    while (ht->occupied[idx] && !ht->deleted[idx] && ht->keys[idx] != key) {
        idx = (idx + step) % ht->size;
    }

    if (!ht->occupied[idx] || ht->deleted[idx]) {
        ht->count++;
    }

    ht->keys[idx] = key;
    ht->values[idx] = value;
    ht->occupied[idx] = true;
    ht->deleted[idx] = false;
}

int double_get(DoubleHashTable* ht, int key, int* found) {
    unsigned int idx = hash_mod(key, ht->size);
    unsigned int step = hash2(key, ht->size);
    int start = idx;

    while (ht->occupied[idx]) {
        if (!ht->deleted[idx] && ht->keys[idx] == key) {
            *found = 1;
            return ht->values[idx];
        }
        idx = (idx + step) % ht->size;
        if (idx == start) break;
    }

    *found = 0;
    return 0;
}

void double_free(DoubleHashTable* ht) {
    free(ht->keys);
    free(ht->values);
    free(ht->occupied);
    free(ht->deleted);
    free(ht);
}

/* =============================================================================
 * 5. Practical: Two Sum
 * ============================================================================= */

int* two_sum(int nums[], int n, int target, int* result_size) {
    ChainHashTable* ht = chain_create(n * 2);
    int* result = malloc(2 * sizeof(int));
    *result_size = 0;

    for (int i = 0; i < n; i++) {
        int complement = target - nums[i];
        int found;
        int idx = chain_get(ht, (char[]){complement + '0', '\0'}, &found);

        if (found) {
            result[0] = idx;
            result[1] = i;
            *result_size = 2;
            chain_free(ht);
            return result;
        }

        /* Simple key conversion (in practice, int should be converted to string) */
        char key[20];
        sprintf(key, "%d", nums[i]);
        chain_insert(ht, key, i);
    }

    chain_free(ht);
    free(result);
    return NULL;
}

/* =============================================================================
 * 6. Practical: Frequency Count
 * ============================================================================= */

void count_frequency(int arr[], int n) {
    LinearHashTable* ht = linear_create(n * 2 + 1);

    for (int i = 0; i < n; i++) {
        int found;
        int count = linear_get(ht, arr[i], &found);
        linear_insert(ht, arr[i], found ? count + 1 : 1);
    }

    printf("    Frequency:\n");
    for (int i = 0; i < ht->size; i++) {
        if (ht->occupied[i] && !ht->deleted[i]) {
            printf("      %d: %d\n", ht->keys[i], ht->values[i]);
        }
    }

    linear_free(ht);
}

/* =============================================================================
 * Test
 * ============================================================================= */

int main(void) {
    printf("============================================================\n");
    printf("Hash Table Examples\n");
    printf("============================================================\n");

    /* 1. Hash Functions */
    printf("\n[1] Hash Functions\n");
    printf("    hash_mod(42, 101) = %u\n", hash_mod(42, TABLE_SIZE));
    printf("    hash_string(\"hello\", 101) = %u\n", hash_string("hello", TABLE_SIZE));
    printf("    hash_string(\"world\", 101) = %u\n", hash_string("world", TABLE_SIZE));

    /* 2. Chaining Hash Table */
    printf("\n[2] Chaining Hash Table\n");
    ChainHashTable* chain_ht = chain_create(TABLE_SIZE);
    chain_insert(chain_ht, "apple", 100);
    chain_insert(chain_ht, "banana", 200);
    chain_insert(chain_ht, "cherry", 300);

    int found;
    printf("    apple: %d\n", chain_get(chain_ht, "apple", &found));
    printf("    banana: %d\n", chain_get(chain_ht, "banana", &found));

    chain_delete(chain_ht, "banana");
    chain_get(chain_ht, "banana", &found);
    printf("    After deleting banana: %s\n", found ? "found" : "not found");
    chain_free(chain_ht);

    /* 3. Linear Probing */
    printf("\n[3] Linear Probing\n");
    LinearHashTable* linear_ht = linear_create(TABLE_SIZE);
    linear_insert(linear_ht, 10, 100);
    linear_insert(linear_ht, 111, 200);  /* Collision: 10 % 101 = 10, 111 % 101 = 10 */
    linear_insert(linear_ht, 212, 300);

    printf("    10: %d\n", linear_get(linear_ht, 10, &found));
    printf("    111: %d\n", linear_get(linear_ht, 111, &found));
    printf("    212: %d\n", linear_get(linear_ht, 212, &found));
    linear_free(linear_ht);

    /* 4. Double Hashing */
    printf("\n[4] Double Hashing\n");
    DoubleHashTable* double_ht = double_create(TABLE_SIZE);
    double_insert(double_ht, 10, 100);
    double_insert(double_ht, 111, 200);
    double_insert(double_ht, 212, 300);

    printf("    10: %d\n", double_get(double_ht, 10, &found));
    printf("    111: %d\n", double_get(double_ht, 111, &found));
    double_free(double_ht);

    /* 5. Frequency Count */
    printf("\n[5] Frequency Count\n");
    int arr[] = {1, 2, 3, 1, 2, 1, 4, 2};
    printf("    Array: [1,2,3,1,2,1,4,2]\n");
    count_frequency(arr, 8);

    /* 6. Hash Table Comparison */
    printf("\n[6] Collision Resolution Comparison\n");
    printf("    | Method       | Pros              | Cons              |\n");
    printf("    |--------------|-------------------|-------------------|\n");
    printf("    | Chaining     | Easy deletion     | Memory overhead   |\n");
    printf("    | Linear Probe | Cache-friendly    | Clustering        |\n");
    printf("    | Double Hash  | Less clustering   | Hash calc cost    |\n");

    printf("\n============================================================\n");

    return 0;
}

File: examples/Algorithm/python/04_hash_table.py
"""
Hash Table
Hash Table Implementation

Implements hash functions and collision resolution methods.
"""

from typing import Any, List, Optional, Tuple
from collections import OrderedDict


# =============================================================================
# 1. Hash Functions
# =============================================================================

def hash_division(key: int, table_size: int) -> int:
    """Division hash function - O(1)"""
    return key % table_size


def hash_multiplication(key: int, table_size: int, A: float = 0.6180339887) -> int:
    """Multiplication hash function (Knuth's suggested A value) - O(1)"""
    return int(table_size * ((key * A) % 1))


def hash_string(s: str, table_size: int) -> int:
    """String hash function (polynomial rolling hash) - O(n)"""
    hash_val = 0
    base = 31
    for char in s:
        hash_val = (hash_val * base + ord(char)) % table_size
    return hash_val


# =============================================================================
# 2. Chaining
# =============================================================================

class HashTableChaining:
    """Hash table with chaining collision resolution"""

    def __init__(self, size: int = 10):
        self.size = size
        self.table: List[List[Tuple[Any, Any]]] = [[] for _ in range(size)]
        self.count = 0

    def _hash(self, key: Any) -> int:
        """Hash function"""
        if isinstance(key, int):
            return key % self.size
        return hash(key) % self.size

    def put(self, key: Any, value: Any) -> None:
        """Insert key-value pair - Average O(1), Worst O(n)"""
        index = self._hash(key)

        # Update existing key
        for i, (k, v) in enumerate(self.table[index]):
            if k == key:
                self.table[index][i] = (key, value)
                return

        # Add new key
        self.table[index].append((key, value))
        self.count += 1

    def get(self, key: Any) -> Optional[Any]:
        """Look up value by key - Average O(1), Worst O(n)"""
        index = self._hash(key)

        for k, v in self.table[index]:
            if k == key:
                return v
        return None

    def remove(self, key: Any) -> bool:
        """Delete key-value pair - Average O(1), Worst O(n)"""
        index = self._hash(key)

        for i, (k, v) in enumerate(self.table[index]):
            if k == key:
                self.table[index].pop(i)
                self.count -= 1
                return True
        return False

    def __contains__(self, key: Any) -> bool:
        return self.get(key) is not None

    def load_factor(self) -> float:
        return self.count / self.size


# =============================================================================
# 3. Open Addressing - Linear Probing
# =============================================================================

class HashTableLinearProbing:
    """Hash table with linear probing collision resolution"""

    DELETED = object()  # Deletion marker

    def __init__(self, size: int = 10):
        self.size = size
        self.keys: List[Any] = [None] * size
        self.values: List[Any] = [None] * size
        self.count = 0

    def _hash(self, key: Any) -> int:
        if isinstance(key, int):
            return key % self.size
        return hash(key) % self.size

    def _probe(self, key: Any, for_insert: bool = False) -> int:
        """Find slot using linear probing"""
        index = self._hash(key)
        first_deleted = -1

        for i in range(self.size):
            probe_index = (index + i) % self.size

            if self.keys[probe_index] is None:
                if for_insert and first_deleted != -1:
                    return first_deleted
                return probe_index

            if self.keys[probe_index] is self.DELETED:
                if first_deleted == -1:
                    first_deleted = probe_index
                continue

            if self.keys[probe_index] == key:
                return probe_index

        return first_deleted if first_deleted != -1 else -1

    def put(self, key: Any, value: Any) -> bool:
        """Insert key-value pair - Average O(1), Worst O(n)"""
        if self.count >= self.size * 0.7:  # Load factor 70%
            self._resize()

        index = self._probe(key, for_insert=True)
        if index == -1:
            return False

        if self.keys[index] is None or self.keys[index] is self.DELETED:
            self.count += 1

        self.keys[index] = key
        self.values[index] = value
        return True

    def get(self, key: Any) -> Optional[Any]:
        """Look up value by key - Average O(1), Worst O(n)"""
        index = self._probe(key)

        if index != -1 and self.keys[index] not in (None, self.DELETED):
            return self.values[index]
        return None

    def remove(self, key: Any) -> bool:
        """Delete key-value pair (using deletion marker) - Average O(1), Worst O(n)"""
        index = self._probe(key)

        if index != -1 and self.keys[index] not in (None, self.DELETED):
            self.keys[index] = self.DELETED
            self.values[index] = None
            self.count -= 1
            return True
        return False

    def _resize(self) -> None:
        """Double the table size"""
        old_keys = self.keys
        old_values = self.values

        self.size *= 2
        self.keys = [None] * self.size
        self.values = [None] * self.size
        self.count = 0

        for k, v in zip(old_keys, old_values):
            if k is not None and k is not self.DELETED:
                self.put(k, v)


# =============================================================================
# 4. Open Addressing - Double Hashing
# =============================================================================

class HashTableDoubleHashing:
    """Hash table with double hashing collision resolution"""

    DELETED = object()

    def __init__(self, size: int = 11):  # Prime number recommended
        self.size = size
        self.keys: List[Any] = [None] * size
        self.values: List[Any] = [None] * size
        self.count = 0

    def _hash1(self, key: Any) -> int:
        if isinstance(key, int):
            return key % self.size
        return hash(key) % self.size

    def _hash2(self, key: Any) -> int:
        """Second hash function (returns non-zero value)"""
        if isinstance(key, int):
            return 7 - (key % 7)  # 7 is a prime smaller than size
        return 7 - (hash(key) % 7)

    def _probe(self, key: Any, for_insert: bool = False) -> int:
        """Find slot using double hashing"""
        h1 = self._hash1(key)
        h2 = self._hash2(key)
        first_deleted = -1

        for i in range(self.size):
            index = (h1 + i * h2) % self.size

            if self.keys[index] is None:
                if for_insert and first_deleted != -1:
                    return first_deleted
                return index

            if self.keys[index] is self.DELETED:
                if first_deleted == -1:
                    first_deleted = index
                continue

            if self.keys[index] == key:
                return index

        return first_deleted if first_deleted != -1 else -1

    def put(self, key: Any, value: Any) -> bool:
        index = self._probe(key, for_insert=True)
        if index == -1:
            return False

        if self.keys[index] is None or self.keys[index] is self.DELETED:
            self.count += 1

        self.keys[index] = key
        self.values[index] = value
        return True

    def get(self, key: Any) -> Optional[Any]:
        index = self._probe(key)
        if index != -1 and self.keys[index] not in (None, self.DELETED):
            return self.values[index]
        return None

    def remove(self, key: Any) -> bool:
        index = self._probe(key)
        if index != -1 and self.keys[index] not in (None, self.DELETED):
            self.keys[index] = self.DELETED
            self.values[index] = None
            self.count -= 1
            return True
        return False


# =============================================================================
# 5. LRU Cache (Least Recently Used Cache)
# =============================================================================

class LRUCache:
    """
    LRU Cache Implementation
    O(1) get/put using OrderedDict
    """

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = OrderedDict()

    def get(self, key: Any) -> Optional[Any]:
        """Look up key and move to most recently used - O(1)"""
        if key not in self.cache:
            return None

        # Move to most recently used
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: Any, value: Any) -> None:
        """Insert key-value pair - O(1)"""
        if key in self.cache:
            self.cache.move_to_end(key)
        else:
            if len(self.cache) >= self.capacity:
                # Remove the oldest entry
                self.cache.popitem(last=False)

        self.cache[key] = value

    def __str__(self) -> str:
        return str(list(self.cache.items()))


# =============================================================================
# 6. Practical Problem: Two Sum
# =============================================================================

def two_sum(nums: List[int], target: int) -> Optional[Tuple[int, int]]:
    """
    Find indices of two numbers that sum to target
    Time Complexity: O(n), Space Complexity: O(n)
    """
    seen = {}  # value -> index

    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return (seen[complement], i)
        seen[num] = i

    return None


# =============================================================================
# 7. Practical Problem: Frequency Count
# =============================================================================

def frequency_count(arr: List[Any]) -> dict:
    """
    Count element frequencies in an array
    Time Complexity: O(n), Space Complexity: O(k) (k = number of unique elements)
    """
    freq = {}
    for item in arr:
        freq[item] = freq.get(item, 0) + 1
    return freq


def most_frequent(arr: List[Any]) -> Optional[Any]:
    """Find the most frequent element"""
    freq = frequency_count(arr)
    if not freq:
        return None
    return max(freq, key=freq.get)


# =============================================================================
# 8. Practical Problem: Subarray Sum Equals K
# =============================================================================

def subarray_sum_count(nums: List[int], k: int) -> int:
    """
    Count contiguous subarrays with sum equal to k
    Uses prefix sum + hash map
    Time Complexity: O(n), Space Complexity: O(n)
    """
    count = 0
    prefix_sum = 0
    prefix_count = {0: 1}  # prefix sum -> occurrence count

    for num in nums:
        prefix_sum += num

        # If prefix_sum - k existed before, the subarray between them has sum k
        if prefix_sum - k in prefix_count:
            count += prefix_count[prefix_sum - k]

        prefix_count[prefix_sum] = prefix_count.get(prefix_sum, 0) + 1

    return count


# =============================================================================
# 9. Practical Problem: Group Anagrams
# =============================================================================

def group_anagrams(strs: List[str]) -> List[List[str]]:
    """
    Group anagrams together
    Time Complexity: O(n * k log k), n=number of strings, k=max string length
    """
    groups = {}

    for s in strs:
        # Use sorted string as key
        key = ''.join(sorted(s))
        if key not in groups:
            groups[key] = []
        groups[key].append(s)

    return list(groups.values())


# =============================================================================
# Tests
# =============================================================================

def main():
    print("=" * 60)
    print("Hash Table Examples")
    print("=" * 60)

    # 1. Hash Function Tests
    print("\n[1] Hash Functions")
    print(f"    hash_division(42, 10) = {hash_division(42, 10)}")
    print(f"    hash_multiplication(42, 10) = {hash_multiplication(42, 10)}")
    print(f"    hash_string('hello', 10) = {hash_string('hello', 10)}")

    # 2. Chaining Test
    print("\n[2] Chaining Method")
    ht_chain = HashTableChaining(5)
    for i, name in enumerate(['Alice', 'Bob', 'Charlie', 'David', 'Eve']):
        ht_chain.put(name, i * 10)
    print(f"    get('Charlie') = {ht_chain.get('Charlie')}")
    print(f"    'Bob' in table = {'Bob' in ht_chain}")
    print(f"    load_factor = {ht_chain.load_factor():.2f}")

    # 3. Linear Probing Test
    print("\n[3] Linear Probing Method")
    ht_linear = HashTableLinearProbing(10)
    for i in range(7):
        ht_linear.put(i * 5, f"value_{i}")
    print(f"    get(10) = {ht_linear.get(10)}")
    ht_linear.remove(10)
    print(f"    get(10) after remove = {ht_linear.get(10)}")

    # 4. LRU Cache Test
    print("\n[4] LRU Cache")
    cache = LRUCache(3)
    cache.put('a', 1)
    cache.put('b', 2)
    cache.put('c', 3)
    print(f"    Cache: {cache}")
    cache.get('a')  # Move 'a' to most recent
    cache.put('d', 4)  # 'b' gets evicted
    print(f"    After get('a'), put('d'): {cache}")

    # 5. Two Sum
    print("\n[5] Two Sum")
    nums = [2, 7, 11, 15]
    target = 9
    result = two_sum(nums, target)
    print(f"    nums = {nums}, target = {target}")
    print(f"    Result: indices {result}")

    # 6. Frequency Count
    print("\n[6] Frequency Count")
    arr = ['a', 'b', 'a', 'c', 'a', 'b']
    freq = frequency_count(arr)
    print(f"    Array: {arr}")
    print(f"    Frequency: {freq}")
    print(f"    Most frequent: {most_frequent(arr)}")

    # 7. Subarray Sum
    print("\n[7] Subarray Sum Equals K Count")
    nums = [1, 1, 1]
    k = 2
    count = subarray_sum_count(nums, k)
    print(f"    nums = {nums}, k = {k}")
    print(f"    Result: {count} subarrays")

    # 8. Group Anagrams
    print("\n[8] Group Anagrams")
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    groups = group_anagrams(strs)
    print(f"    Input: {strs}")
    print(f"    Groups: {groups}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()
